map/bookmark_manager.cpp

```diff
@@ -404,6 +404,59 @@ void BookmarkManager::ResetRecentlyDeletedBookmark()
   m_recentlyDeletedBookmark.reset();
 }
 
+size_t BookmarkManager::GetRecentlyDeletedCategoriesCount() const
+{
+  Platform::FilesList files;
+  Platform::GetFilesByExt(GetTrashDirectory(), kKmlExtension, files);
+  return files.size();
+}
+
+BookmarkManager::KMLDataCollectionPtr BookmarkManager::GetRecentlyDeletedCategories()
+{
+  auto collection = LoadBookmarks(GetTrashDirectory(), kKmlExtension, KmlFileType::Text,
+                                  [](kml::FileData const &)
+                                  {
+    return true;
+  });
+  return collection;
+}
+
+bool BookmarkManager::IsRecentlyDeletedCategory(const std::string &filePath) const
+{
+  Platform::FilesList files;
+  Platform::GetFilesByExt(GetTrashDirectory(), kKmlExtension, files);
+  auto const it = std::find(files.begin(), files.end(), base::FileNameFromFullPath(filePath));
+  return it != files.end();
+}
+
+void BookmarkManager::RecoverRecentlyDeletedCategoriesAtPaths(std::vector<std::string> const & filePaths)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
+  GetPlatform().RunTask(Platform::Thread::File, [this, filePaths]()
+  {
+    for (auto const & deletedFilePath : filePaths)
+    {
+      CHECK(IsRecentlyDeletedCategory(deletedFilePath), ("The category at path", deletedFilePath, "should be in the trash."));
+      CHECK(Platform::IsFileExistsByFullPath(deletedFilePath), ("File should exist to be recovered.", deletedFilePath));
+      auto recoveredFilePath = GenerateValidAndUniqueFilePathForKML(base::GetNameFromFullPathWithoutExt(deletedFilePath));
+      base::MoveFileX(deletedFilePath, recoveredFilePath);
+      GetPlatform().RunTask(Platform::Thread::Gui, [this, recoveredFilePath]() {
+        ReloadBookmark(recoveredFilePath);
+      });
+    }
+  });
+}
+
+void BookmarkManager::DeleteRecentlyDeletedCategoriesAtPaths(std::vector<std::string> const & deletedFilePaths)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
```

Comment:
This call may contain a lot of file delete operations. Would it be better to remove the thread checker line and run it on the async File thread instead of the UI thread? It doesn't return anything. But I assume that the caller expects that after returning to the main bookmarks screen files are already deleted, and an updated count in Recently Deleted is displayed. Correct?

Reply:
The main idea is to call this method synchronously on the main and Not pass the completion handler -> update the state immediately without waiting for deletion finish.
Deletion is quite fast and it is better to rely on the actual directory content.

For example:
1. user press `delete all`
2. deletion task is dispatched to the file thread
3. the screen is closed
4. the code checks the trash directory and may see that not all the files id deleted - if will cause a bug

Reply:
I understand the reasoning for the synchronous call.

Won't the android linter/runtime checker complain for file operations made on the UI thread in this case?